File: src/main/java/at/hannibal2/skyhanni/config/commands/SimpleCommand.java
package at.hannibal2.skyhanni.config.commands;

import at.hannibal2.skyhanni.test.command.ErrorManager;
import net.minecraft.command.CommandBase;
import net.minecraft.command.ICommandSender;
import net.minecraft.util.BlockPos;

import java.util.List;

/**
 * @author Moulberry
 **/
public class SimpleCommand extends CommandBase {

    private final String commandName;
    private final ProcessCommandRunnable runnable;
    private TabCompleteRunnable tabRunnable;

    public SimpleCommand(String commandName, ProcessCommandRunnable runnable) {
        this.commandName = commandName;
        this.runnable = runnable;
    }

    public SimpleCommand(String commandName, ProcessCommandRunnable runnable, TabCompleteRunnable tabRunnable) {
        this.commandName = commandName;
        this.runnable = runnable;
        this.tabRunnable = tabRunnable;
    }

    public abstract static class ProcessCommandRunnable {

        public abstract void processCommand(ICommandSender sender, String[] args);
    }

    public interface TabCompleteRunnable {

         List<String> tabComplete(ICommandSender sender, String[] args, BlockPos pos);
    }

    public boolean canCommandSenderUseCommand(ICommandSender sender) {
        return true;
    }

    public String getCommandName() {
        return commandName;
    }

    public String getCommandUsage(ICommandSender sender) {
        return "/" + commandName;
    }

    @Override
    public void processCommand(ICommandSender sender, String[] args) {
        try {
            runnable.processCommand(sender, args);
        } catch (Throwable e) {
            ErrorManager.INSTANCE.logError(e, "Error while running command /" + commandName);
        }
    }

    @Override
    public List<String> addTabCompletionOptions(ICommandSender sender, String[] args, BlockPos pos) {
        if (tabRunnable != null) return tabRunnable.tabComplete(sender, args, pos);
        return null;
    }
}

File: src/main/java/at/hannibal2/skyhanni/config/commands/SimpleCommand.kt
package at.hannibal2.skyhanni.config.commands

import at.hannibal2.skyhanni.test.command.ErrorManager
import net.minecraft.command.CommandBase
import net.minecraft.command.ICommandSender
import net.minecraft.util.BlockPos

class SimpleCommand(
    private val name: String,
    private val aliases: List<String>,
    private val callback: (Array<String>) -> Unit,
    private val tabCallback: ((Array<String>) -> List<String>) = { emptyList() },
) : CommandBase() {

    override fun canCommandSenderUseCommand(sender: ICommandSender) = true
    override fun getCommandName() = name
    override fun getCommandAliases() = aliases
    override fun getCommandUsage(sender: ICommandSender) = "/$name"

    override fun processCommand(sender: ICommandSender, args: Array<String>) {
        try {
            callback(args)
        } catch (e: Throwable) {
            ErrorManager.logErrorWithData(e, "Error while running command /$name")
        }
    }

    override fun addTabCompletionOptions(sender: ICommandSender, args: Array<String>, pos: BlockPos) =
        tabCallback(args).takeIf { it.isNotEmpty() }
}
